datafusion/optimizer/src/reduce_cross_join.rs

```diff
@@ -0,0 +1,442 @@
+
+        reduce_cross_join(self, plan, &mut possible_join_keys, &mut all_join_keys, optimizer_config)
+    }
+
+    fn name(&self) -> &str {
+        "reduce_cross_join"
+    }
+}
+
+/// Attempt to reduce cross joins to inner joins.
+/// for queries: select ... from a, b where a.x = b.y and b.xx = 100;
+/// select ... from a, b where (a.x = b.y and b.xx = 100) or (a.x = b.y and b.xx = 200);
+/// select ... from a, b, c where (a.x = b.y and b.xx = 100 and a.z = c.z)
+/// or (a.x = b.y and b.xx = 200 and a.z=c.z);
+/// For above queries, the join predicate is available in filters and they are moved to
+/// join nodes appropriately
+/// This fix helps to improve the performance of TPCH Q19. issue#78
+///
+fn reduce_cross_join(
+    _optimizer: &ReduceCrossJoin,
+    plan: &LogicalPlan,
+    possible_join_keys: &mut Vec<(Column, Column)>,
+    all_join_keys: &mut HashSet<(Column, Column)>,
+    _optimizer_config: &OptimizerConfig,
+) -> Result<LogicalPlan> {
+        match plan {
+            LogicalPlan::Filter(Filter { input, predicate }) => {
+                extract_possible_join_keys(predicate, possible_join_keys)?;
+                let left = reduce_cross_join(_optimizer, input, possible_join_keys, all_join_keys, _optimizer_config)?;
+
+                // if there are no join keys then do nothing.
+                if all_join_keys.is_empty() {
+                    Ok(plan.clone())
+                }
+                else {
+                    // remove join expressions from filter
+                    match remove_join_expressions(predicate, all_join_keys)? {
+                        Some(filter_expr) => {
+                            Ok(LogicalPlan::Filter(Filter {
+                                predicate: filter_expr,
+                                input: Arc::new(left),
+                            }))
+                        }
+                        _ => Ok(left)
+                    }
+                }
+            }
+            LogicalPlan::CrossJoin(cross_join) => {
+                let left_plan = reduce_cross_join(
+                    _optimizer,
+                    &cross_join.left,
+                    possible_join_keys,
+                    all_join_keys,
+                    _optimizer_config,
+                )?;
+                let right_plan = reduce_cross_join(
+                    _optimizer,
+                    &cross_join.right,
+                    possible_join_keys,
+                    all_join_keys,
+                    _optimizer_config,
+                )?;
+                // can we find a match?
+                let left_schema = left_plan.schema();
+                let right_schema = right_plan.schema();
+                let mut join_keys = vec![];
+
+                for (l, r) in possible_join_keys {
+                    if left_schema.field_from_column(l).is_ok()
+                        && right_schema.field_from_column(r).is_ok()
+                        && can_hash(
```

Comment:
Nice! I had no idea this function existed, and would have used it if I knew it did :)